### PR TITLE
[POC][WIP] a new test for detecting ungraceful kube-apiserver termination by kubelet

### PR DESCRIPTION
**Pull request by p0lyn0mial**
Initially I wanted to compare shutdown related events but after [inspecting](https://prow.ci.openshift.org/view/gs/test-platform-results/logs/periodic-ci-openshift-release-master-ci-4.18-e2e-gcp-ovn-upgrade/1822964770265894912) I have realised that event's are sometimes not even persisted in the storage.

The new test added by this PR simply looks for the `Previous pod .* did not terminate gracefully string in the -termination.log` files.

This PR could be made simpler and smaller by copying the existing code instead of reusing it.

### CODE REVIEW

File: test/extended/apiserver/graceful_termination.go [hunk 1]
```diff
@@ -61,6 +62,45 @@ var _ = g.Describe("[sig-api-machinery][Feature:APIServer][Late]", func() {
...
+			g.Skip("oc must-gather doesn't currently support external controlPlaneTopology")
+		}
+
+		tempDir, err := ioutil.TempDir("", "test.oc-adm-must-gather.")
+		o.Expect(err).NotTo(o.HaveOccurred())
+		defer os.RemoveAll(tempDir)
+
+		var finalMessageBuilder strings.Builder
+		terminationRegexp := regexp.MustCompile(`Previous pod .* did not terminate gracefully`)
+		terminationLogsPerServer := gatherMustGatherFor(oc, tempDir, clihelpers.IsTerminationLog, extractAPIServerNameFromTerminationFile)
```
Comment (deads2k) [inline test/extended/apiserver/graceful_termination.go:79]: Don't use must-gather for this.  If you just need a hostmount pod, use that.  Removing layers will help future us.

File: test/extended/apiserver/graceful_termination.go [hunk 2]
```diff
@@ -61,6 +62,45 @@ var _ = g.Describe("[sig-api-machinery][Feature:APIServer][Late]", func() {
...
+		if *controlPlaneTopology == configv1.ExternalTopologyMode {
+			g.Skip("oc must-gather doesn't currently support external controlPlaneTopology")
+		}
+
+		tempDir, err := ioutil.TempDir("", "test.oc-adm-must-gather.")
+		o.Expect(err).NotTo(o.HaveOccurred())
+		defer os.RemoveAll(tempDir)
+
+		var finalMessageBuilder strings.Builder
+		terminationRegexp := regexp.MustCompile(`Previous pod .* did not terminate gracefully`)
```
Comment (deads2k) [inline test/extended/apiserver/graceful_termination.go:78]: Why doesn't the  kube-apiserver itself produce an event when this is detected?

File: test/extended/apiserver/graceful_termination.go (hunk 1 above)
Comment (p0lyn0mial) [reply test/extended/apiserver/graceful_termination.go:79]: What is your concern here? 

Please note that this test doesn't download the entire must-gather tarball. It uses must-gather to only collect the audit logs. While it is true that the tarball contains more files than this test needs, I should have given the function a more specific name.

An alternative approach would be to copy only the files we need. Starting new pods on each master, checking their state, and then copying the output doesn't seem simpler.

Another option could be to use `oc adm node-logs` to copy specific files. However, we would still need to identify all the nodes on which this command should be executed.

File: test/extended/apiserver/graceful_termination.go (hunk 2 above)
Comment (p0lyn0mial) [reply test/extended/apiserver/graceful_termination.go:78]: It [does](https://github.com/openshift/kubernetes/blob/6892b57d65d25fb0588693bce3d338d8b8b1d2b4/cmd/watch-termination/main.go#L149), we even have a test for [it](https://github.com/openshift/origin/blob/master/test/extended/apiserver/graceful_termination.go#L53). The issue is that sometimes the events are not reported (probably deserves more investigation). For example, the test passed in [the faulty run](https://prow.ci.openshift.org/view/gs/test-platform-results/logs/periodic-ci-openshift-release-master-ci-4.18-e2e-gcp-ovn-upgrade/1822964770265894912) even though the servers were not gracefully terminated.

File: test/extended/apiserver/graceful_termination.go (hunk 1 above)
Comment (p0lyn0mial) [reply test/extended/apiserver/graceful_termination.go:79]: I have opened a PR that doesn't use the must-gather tool for downloading audit logs - https://github.com/openshift/origin/pull/29015

**Comment by p0lyn0mial** [comment]
/close 

in favour of https://github.com/openshift/origin/pull/29015

**Pull request closed without merging**